sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/codegen/CodeGenerator.scala

```diff
@@ -276,6 +276,11 @@ class CodegenContext {
       case t: DecimalType => s"$row.setDecimal($ordinal, $value, ${t.precision})"
       // The UTF8String may came from UnsafeRow, otherwise clone is cheap (re-use the bytes)
       case StringType => s"$row.update($ordinal, $value.clone())"
+      // InternalRow, ArrayData and MapData may came from UnsafeRow, we should copy it to avoid
+      // keeping a "pointer" to a memory region which is out of control from the updated row.
+      case _: StructType => s"$row.update($ordinal, $value.copy())"
```

Comment:
If `$row` could do the copy inside `update`, then we do need to do the copy here, right?

Maybe it's time to check all the MutableRow, MutableProjection, to see where is the best place to do the copy.